### PR TITLE
[SYCL] Fix operators for `bool` swizzled vec

### LINKED ISSUE #11995: `operator&&` result differs for certain swizzle operands
**Describe the bug**
The `operator&&` gives a different result for   `testVec1.template swizzle<sycl::elem::s0>() && testVec2` and `testVec1 && testVec2`.

**To Reproduce**
```c++
#include <sycl/sycl.hpp>

template <typename T, typename ResultT> void testAndOperator() {
  constexpr int N = 4;
  std::array<ResultT, N> test_results{};

  sycl::queue q;
  sycl::buffer<ResultT, 1> buffer{test_results.data(), N};
  q.submit([&](sycl::handler &cgh) {
     auto acc = buffer.template get_access<sycl::access_mode::write>(cgh);
     cgh.parallel_for(sycl::range<1>{1}, [=](sycl::id<1> id) {
       auto testVec1 = sycl::vec<T, 1>(static_cast<T>(1));
       auto testVec2 = sycl::vec<T, 1>(static_cast<T>(2));
       sycl::vec<ResultT, 1> resVec;

       ResultT expected = static_cast<ResultT>(
           -(static_cast<ResultT>(1) && static_cast<ResultT>(2)));
       acc[0] = expected;

       // LHS swizzle
       resVec = testVec1.template swizzle<sycl::elem::s0>() && testVec2;
       acc[1] = resVec[0];

       // RHS swizzle
       resVec = testVec1 && testVec2.template swizzle<sycl::elem::s0>();
       acc[2] = resVec[0];

       // No swizzle
       resVec = testVec1 && testVec2;
       acc[3] = resVec[0];
     });
   }).wait();

  for (int i = 0; i < N; i++) {
    std::cout << "Test " << i << ": " << ((int)test_results[i]) << std::endl;
  }
  std::cout << std::endl;
}

int main() {
  std::cout << "Testing with T = bool" << std::endl;
  testAndOperator<bool, std::int8_t>();
  std::cout << "Testing with T = std::int8_t" << std::endl;
  testAndOperator<std::int8_t, std::int8_t>();
}
```
```
$ clang++ -fsycl ./reproducer.cpp
$ ./a.out
Testing with T = bool
Test 0: -1
Test 1: 1
Test 2: -1
Test 3: -1

Testing with T = std::int8_t
Test 0: -1
Test 1: -1
Test 2: -1
Test 3: -1
```
The Test 1 result for type `bool` differs from the others. This is a bug since the three test expressions should produce the same result.

**Environment (please complete the following information):**

- OS: Linux
- Target device and vendor: Intel CPU
- DPC++ version: 6bce7f64f51a4370052bffa3fa257ca16d8aad9e
- Dependencies version: OpenCL [2023.16.6.0.28_042959]

**Additional context**
I'm not sure if the correct test result is `-1` or `1`. The SYCL specification states the following:

> Construct a new instance of the SYCL `vec` class template with the same template parameters as `lhs vec` with each element of the new SYCL `vec` instance the result of an element-wise `OP` logical operation between each element of `lhs vec` and each element of the `rhs` SYCL `vec`.

The result of the element-wise `&&` is `true`. Converting the result to an `std::int8_t` in C++ results in `1`, but SYCL states the following for other operators with a logical result:

> Each element of the SYCL `vec` that is returned must be `-1` if the operation results in `true` and `0` if the operation results in `false`.

### PR DESCRIPTION
**Pull request by 0x12CC**
c3a9615441c86172364357b076decdfcb598cb5c added `CommonDataT` to fix an issue where results were truncated if an operand had a larger type than the result. As `CommonDataT` uses `std::common_type_t` of the left and right operands, it ignores `DataT`. This causes an issue for cases where `CommonDataT = bool`  and `DataT != bool` since the operation result will be implicitly converted from `DataT` to `bool`.

This change adds `DataT` to set of types used to define `CommonDataT`. Cases where the operands were `bool` type will now have the correct result rather than `1`. Fixes #11995.

### CODE REVIEW

File: sycl/include/sycl/types.hpp [hunk 1]
```diff
@@ -1544,8 +1544,16 @@ template <typename VecT, typename OperationLeftT, typename OperationRightT,
...
+  using CommonDataT = std::conditional_t<
+      sizeof(DataT) >=
+          sizeof(std::common_type_t<typename OperationLeftT::DataT,
+                                    typename OperationRightT::DataT>),
+      DataT,
+      std::common_type_t<typename OperationLeftT::DataT,
+                         typename OperationRightT::DataT>>;
+
+  //  std::common_type_t<DataT, typename OperationLeftT::DataT,
+  //                                        typename OperationRightT::DataT>;
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1556]: Development/debug leftover.

File: sycl/include/sycl/types.hpp [hunk 2]
```diff
@@ -1544,8 +1544,16 @@ template <typename VecT, typename OperationLeftT, typename OperationRightT,
           template <typename> class OperationCurrentT, int... Indexes>
 class SwizzleOp {
   using DataT = typename VecT::element_type;
-  using CommonDataT = std::common_type_t<typename OperationLeftT::DataT,
-                                         typename OperationRightT::DataT>;
+  using CommonDataT = std::conditional_t<
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1547]: I'd add `using OpLeftDataT`/`OpRightDataT` here, maybe with different names. Also, we need a comment explaing what we're doing and **why**.

File: sycl/test-e2e/Regression/vec_rel_swizzle_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+template <typename T, typename ResultT>
+bool testAndOperator(const std::string &typeName) {
+  constexpr int N = 5;
+  std::array<ResultT, N> results{};
+
+  sycl::queue q;
+  sycl::buffer<ResultT, 1> buffer{results.data(), N};
+  q.submit([&](sycl::handler &cgh) {
+     auto acc = buffer.template get_access<sycl::access_mode::write>(cgh);
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Regression/vec_rel_swizzle_ops.cpp:18]:
```suggestion
     sycl::accessor acc{buffer, cgh, sycl::write_only};
```

File: sycl/include/sycl/types.hpp [hunk 4]
```diff
@@ -1544,8 +1544,14 @@ template <typename VecT, typename OperationLeftT, typename OperationRightT,
           template <typename> class OperationCurrentT, int... Indexes>
 class SwizzleOp {
   using DataT = typename VecT::element_type;
-  using CommonDataT = std::common_type_t<typename OperationLeftT::DataT,
-                                         typename OperationRightT::DataT>;
+  // Certain operators return a vector with a different element type. Also, the
+  // left and right operand types may differ. CommonDataT selects a result type
+  // based on these types to ensure that the result value can be represented.
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1549]: Honestly, I wanted something more concrete/specific here, maybe with a few example to represent each scenario. The cases when `std::common_type_t` fails are of particular interest.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by sarnex** [comment]
@0x12CC It looks like `vec_rel_swizzle_ops.cpp` is failing, I see it failing in my PR [here](https://github.com/intel/llvm/pull/12029)

**Comment by sarnex** [comment]
Cool thx

**Comment by steffenlarsen** [comment]
Failure should be fixed with https://github.com/intel/llvm/pull/12040.

**Pull request merged**